File: Problem5.java
import java.util.function.BiFunction;
import java.util.function.Function;

// This problem was asked by Jane Street.

// cons(a, b) constructs a pair, and car(pair) and cdr(pair) returns the first and last element of that pair.
// For example, car(cons(3, 4)) returns 3, and cdr(cons(3, 4)) returns 4.


// Given this implementation of cons:

// def cons(a, b):
//     def pair(f):
//         return f(a, b)
//     return pair
// Implement car and cdr.

//##########################################################################


// Trying it in Java, because life is suffering

class Pair {
    final int a;
    final int b;

    public Pair(int a, int b) {
        this.a = a;
        this.b = b;
    }
}

public class Problem5 {

    Function<BiFunction<Integer, Integer, Pair>, Pair> cons(Integer a, Integer b) {
        return (BiFunction<Integer, Integer, Pair> c) -> (Pair) c.apply(a,b);
    }

    Pair f(Integer a , Integer b){
        return new Pair(a,b);
    }

    int car(Function<BiFunction<Integer, Integer, Pair>, Pair> supplier) {
        return supplier.apply(this::f).a;
    }

    int cdr(Function<BiFunction<Integer, Integer, Pair>, Pair> supplier) {
        return supplier.apply(this::f).b;
    }

    
    public static void main(String[] args) {
        Problem5 problem = new Problem5();

        assert 3 == problem.car(problem.cons(3,4));
        assert 4 == problem.cdr(problem.cons(3,4));
    }
}

File: problem5.py
# This problem was asked by Jane Street.

# cons(a, b) constructs a pair, and car(pair) and cdr(pair) returns the first and last element of that pair. 
# For example, car(cons(3, 4)) returns 3, and cdr(cons(3, 4)) returns 4.

# Given this implementation of cons:

# def cons(a, b):
#     def pair(f):
#         return f(a, b)
#     return pair
# Implement car and cdr.

##########################################################################

#Thats more of a tricky python question than programming in general


# This is the tricky function we receive:
def cons(a, b):
    def pair(f):        # 'cons' creates another function, 'pair', wich receives a parameter 'f'
        return f(a, b)  # Parameter 'f' turns out to be yet another function, that is not provided!
    return pair         # In the end, what is returned is not a proper pair, its a 
                        # function, that receives a (inexistent) function wich must create a pair...

# Then you have to code this sucker here
def f(a,b):
    return (a,b) 

# In the two required functions, you have to call 'pair', passing 'f' as parameter
# and then you get your pair...
def car(pair):
    return pair(f)[0]

def cdr(pair):
    return pair(f)[1]


# TESTS

assert 3 == car(cons(3, 4))
assert 4 == cdr(cons(3, 4))
